nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,573 @@
+  const double lookahead_dist = getLookAheadDistance(speed);
+  auto carrot_pose = getLookAheadMarker(lookahead_dist, transformed_plan);
+  carrot_pub_->publish(std::move(createCarrotMsg(carrot_pose)));
+
+  double linear_vel, angular_vel;
+
+  // Find distance^2 to look ahead point (carrot) in robot base frame
+  // This is the chord length of the circle
+  const double carrot_dist2 =
+    (carrot_pose.pose.position.x * carrot_pose.pose.position.x) +
+    (carrot_pose.pose.position.y * carrot_pose.pose.position.y);
+
+  // Find curvature of circle (k = 1 / R)
+  double curvature = 0.0;
+  if (carrot_dist2 > 0.001) {
+    curvature = 2.0 * carrot_pose.pose.position.y / carrot_dist2;
+  }
+
+  // Apply curvature to angular velocity
+  linear_vel = desired_linear_vel_;
+  angular_vel = desired_linear_vel_ * curvature;
+
+  // Make sure we're in compliance with basic constraints
+  double angle_to_path;
+  if (shouldRotateToPath(carrot_pose, angle_to_path)) {
+    rotateToHeading(linear_vel, angular_vel, angle_to_path, speed);
+  } else {
+    applyConstraints(
+      linear_vel, fabs(lookahead_dist - sqrt(carrot_dist2)),
+      lookahead_dist, curvature, speed,
+      costAtPose(pose.pose.position.x, pose.pose.position.y));
+  }
+
+  // Collision checking on this velocity heading
+  if (isCollisionImminent(pose, carrot_pose, curvature, linear_vel, angular_vel)) {
+    RCLCPP_ERROR(logger_, "Collision imminent!");
+    throw std::runtime_error("RegulatedPurePursuitController detected collision ahead!");
+  }
+
+  // populate and return message
+  geometry_msgs::msg::TwistStamped cmd_vel;
+  cmd_vel.header = pose.header;
+  cmd_vel.twist.linear.x = linear_vel;
+  cmd_vel.twist.angular.z = angular_vel;
+  return cmd_vel;
+}
+
+bool RegulatedPurePursuitController::shouldRotateToPath(
+  const geometry_msgs::msg::PoseStamped & carrot_pose, double & angle_to_path)
+{
+  // Whether we should rotate robot to rough path heading
+  angle_to_path = atan2(carrot_pose.pose.position.y, carrot_pose.pose.position.x);
+  if (use_rotate_to_heading_ && fabs(angle_to_path) > rotate_to_heading_min_angle_) {
+    return true;
+  }
+
+  return false;
+}
+
+void RegulatedPurePursuitController::rotateToHeading(
+  double & linear_vel, double & angular_vel,
+  const double & angle_to_path, const geometry_msgs::msg::Twist & curr_speed)
+{
+  // Rotate in place using max angular velocity / acceleration possible
+  linear_vel = 0.0;
+  const double sign = angle_to_path > 0.0 ? 1.0 : -1.0;
+  angular_vel = sign * rotate_to_heading_angular_vel_;
+
+  const double & dt = control_duration_;
+  const double min_feasible_angular_speed = curr_speed.angular.z - max_angular_accel_ * dt;
+  const double max_feasible_angular_speed = curr_speed.angular.z + max_angular_accel_ * dt;
+  angular_vel = std::clamp(angular_vel, min_feasible_angular_speed, max_feasible_angular_speed);
+}
+
+geometry_msgs::msg::PoseStamped RegulatedPurePursuitController::getLookAheadMarker(
```

Comment:
```suggestion
geometry_msgs::msg::PoseStamped RegulatedPurePursuitController::getLookAheadPose(
```
Marker might be misleading as visualization marker.